### ISSUE TITLE
i40e: frames received in blocks of 4 buffers on some cards

### ISSUE THREAD

**Issue by desbma** [ntop/PF_RING#244]
I am trying various software configurations to capture with 0% loss, at full link speed on 2 x 10 GbE interfaces.
I am doing my tests with 60 B frames to simulate worst case.

I tried the following, everytime with some percentage of lost frames:
* 2 devices, 1 balancer, 2 reader threads (`zbalance`) => 15-20% loss
* 2 devices, 1 balancer per device, and a reader thread per balancer => <0.1% loss
  This configuration segfaults when compiled in debug mode, see #242, so there may be some silent internal PF_Ring data corruption even when compiling with `-O2`.
* 2 devices, a reader thread per device => 1-5 frames lost over the millions I have sent

The CPU hosting the 2 reader threads are only loaded by ~35% in the last test.
I tried playing with the burst length, number of threads, etc. but I never managed to get 0 loss, even though it's pretty close in the last case.

The only thing I have not yet tested is to have 2 RSS queues per interface (so 4 PF_Ring devices), and a reader thread per device, but that last test with only a few frames lost over several millions makes me think I should be able to get 0.

My environment is the following:

* PF_Ring ZC 6.6.0 (commercial license)
* Linux kernel 4.4 (from Ubuntu 16.04)
* 2 x 10GbE Intel XL710 NICs
* 16 GB of 2MB hugepages
* MTU set to 9000

Thank you

**Comment by cardigliano** [comment]
@desbma what did you use in the last test? 
Please use zcount and report us 1. command line, 2. output and 3. top output (pressing 1 to see all cores load)
BTW, please note ixgbe cards (e.g. X520) have much more buffering power than i40e cards (e.g. X710), thus I recommend you to test X520 if you are still in the process of selecting the NIC.

**Comment by desbma** [comment]
> @desbma what did you use in the last test?

A program I wrote. I can post it if that helps.

> Please use zcount and report us 1. command line, 2. output and 3. top output (pressing 1 to see all cores load)

`zcount` can only open a single interface per invocation so this is irrelevant for this issue, which is specific to 2 interfaces. With a single interface I can easily get 0 loss.
I can do the test with `zbalance` if that helps (first configuration in the list of my first post), however in that case the single balancer becomes the bottleneck and performance drops.

> BTW, please note ixgbe cards (e.g. X520) have much more buffering power than i40e cards (e.g. X710), thus I recommend you to test X520 if you are still in the process of selecting the NIC.

Thanks for the suggestion however in my case the hardware is already chosen, only the capture framework is not. I'm currently in the process of evaluating Netmap, PF_Ring, DPDK, etc.
Also I don't need buffering power in the NIC, my CPUs are a lot more powerful than what is needed to capture packets at line speed, as long as no copy, allocation or costly processing occurs on the fast path.

**Comment by cardigliano** [comment]
> zcount can only open a single interface per invocation so this is irrelevant for this issue, which is specific to 2 interfaces. With a single interface I can easily get 0 loss.

Please run 2 intances at the same time, binding them to different cores, you should get the same performance of a multithreaded application

**Comment by desbma** [comment]
Second `zcount` invocation fails with `pfring_zc_create_cluster error [Invalid argument] Please check that pf_ring.ko is loaded and hugetlb fs is mounted`.

I tried reducing `MAX_CARD_SLOTS` to 16384 and then 8192, same error.

**Comment by cardigliano** [comment]
Please set a different cluster id (-c option)

**Comment by desbma** [comment]
Thanks, so with 2 invocations I have 0 loss every time.

The only difference I see with my program is that I use a single cluster for both interfaces (because of #243).

I will try again when that other one is fixed.

**Comment by desbma** [comment]
I did more testing, here is what I found out:

In my test environment I generate traffic with Linux's pktgen module. 
The number of frames sent can vary a bit, so at the end of tests I query the number of frames sent by pktgen for each interface, and compare with what the capturing program have received.

If the tx count is a nice multiple of 16 like 14 880 960 frames for 1 second of traffic, I get no loss with my program or 2 `zcount` instances.
However if that number is a little off like 14 880 9768 or 14 880 967, I get a few lost frames, sometimes on an interface and not the other.

Contrary to what I wrote before, **this is reproducible with both my program or a pair of `zcount` instances**.

At first I though my test environment was sending spurious frames not counted by pktgen, however:
* I checked with tcpdump on the receiving side, and no frames is send other than the pktgen traffic
* Using the same exact test procedure, with Netmap, I capture 100% of frames (rx count == tx count) even if the tx counter varies a bit

I tried adding a delay between the program startup and the start of traffic, and also between the end of traffic and the time I query counters on the rx side, but that did not fix it.
Also I tried with 1s or 30s of traffic and the number of lost frames is still very low (usually 2 or 3).

So I think this may be an issue with the way PF_Ring handles the last incoming packets for packet counts not multiple of 16 (or 8 ?).

As requested previously, here is the output of the zcount invocations, in that case pktgen sent 14 880 978 frames on each interface:
```
 $ sudo ~/PF_RING-6.6.0/userland/examples_zc/zcount -i zc:enp15s0f0 -c 0 -g 4
=========================
Absolute Stats: 0 pkts (0 drops) - 0 bytes
=========================

=========================
Absolute Stats: 0 pkts (0 drops) - 0 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

=========================
Absolute Stats: 4'455'521 pkts (0 drops) - 374'263'764 bytes
Actual Stats: 4'455'106.67 pps (0.00 drops) - 2.99 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 10'424'422.98 pps (0.00 drops) - 7.01 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

^CLeaving...
=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================
```

```
$ sudo ~/PF_RING-6.6.0/userland/examples_zc/zcount -i zc:enp15s0f1 -c 1 -g 8
=========================
Absolute Stats: 0 pkts (0 drops) - 0 bytes
=========================

=========================
Absolute Stats: 0 pkts (0 drops) - 0 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

=========================
Absolute Stats: 14'488'312 pkts (0 drops) - 1'217'018'208 bytes
Actual Stats: 14'486'935.74 pps (0.00 drops) - 9.74 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 392'620.81 pps (0.00 drops) - 0.26 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

^CLeaving...
=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================

=========================
Absolute Stats: 14'880'976 pkts (0 drops) - 1'250'001'984 bytes
Actual Stats: 0.00 pps (0.00 drops) - 0.00 Gbps
=========================
```

I can't post the output of `top` because I have too much cores (4 nodes of 14 physical cores each, so 112 total with hyper threading) so they don't fit in the screen, but the load is about 35-40% of user time on cores 4 and 8, the rest is idle.

**Comment by desbma** [comment]
I did further analysis, and I can confirm the problem is due to the last few frames which are lost **when the frame count is not multiple of 4**. I have renamed the issue accordingly.

To reproduce very easily, with systematic occurence:

1. Start `zcount` with for example:
`zcount -i zc:enp15s0f0 -c 0 -g 4`

2. Send 4 frames

3. See `zcount` counter if correctly incremented by 4

4. Send 3 frames

5. **See `zcount` counter does not change, and stops updating even when waiting several seconds**

6. Send 1 frame, see counter updates to 8
